### PR TITLE
feat(sdk): Add `LinkedChunk::subscribe_as_vector`

### PR DESCRIPTION
**Pull request by Hywan**
This patch implements `LinkedChunk::<Item, Gap>::as_vector`,
which returns a `AsVector`.

`AsVector` is a type that transforms a
`Vec<Update<Item, Gap>>` into a `Vec<VectorDiff<Item>>` —this type—. Basically, it
helps to consume a `LinkedChunk<CAP, Item, Gap>` as if it was an
`eyeball::ObservableVector<Item>`.

How this type transforms `Update` into `VectorDiff`? There is no internal
buffer of kind `eyeball_im::ObservableVector<Item>`, which could have been
used to generate the `VectorDiff`s. They are computed manually.

The only buffered data is pairs of `ChunkIdentifier` and `ChunkLength`.
The following rules must be respected:

* A chunk of kind `ChunkContent::Gap` has a length of 0,
* A chunk of kind `ChunkContent::Items` has a length equals to its number
  of items,
* The pairs must be ordered exactly like the chunks in `LinkedChunk`, i.e.
  the first pair must represent the first chunk, the last pair must
  represent the last chunk.

The only thing this algorithm does is maintaining the pairs:

* `Update::NewItemsChunk` and `Update::NewGapChunk` are inserting a new
  pair with a chunk length of 0 at the appropriate index,
* `Update::RemoveChunk` is removing a pair,
* `Update::PushItems` is increasing the length of the appropriate pair by
  the number of new items, and is potentially emitting `VectorDiff`,
* `Update::DetachLastItems` is decreasing the length of the appropriate pair
  by the number of items to be detached; no `VectorDiff` is emitted,
* `Update::StartReattachItems` and `Update::EndReattachItems` are
  respectively muting or unmuting the emission of `VectorDiff` by
  `Update::PushItems`.

The only `VectorDiff` that are emitted are `VectorDiff::Insert` or
`VectorDiff::Append` because a `LinkedChunk` is append-only.

`VectorDiff::Append` is an optimisation when numerous `VectorDiff::Insert`s
have to be emitted at the last position.

`VectorDiff::Insert` need an index. To compute this index, the algorithm
will iterate over all pairs to accumulate each chunk length until it finds
the appropriate pair (given by `Update::PushItems::position_hint`). This
is _the offset_. To this offset, the algorithm adds the position's index of
the new items (still given by `Update::PushItems::position_hint`). This is
_the index_. This logic works for all cases as long as pairs are maintained
according to the rules hereinabove.

That's a pretty memory compact and computation efficient way to map a
`Vec<Update<Item, Gap>>` into a `Vec<VectorDiff<Item>>`. The larger
the `LinkedChunk` capacity is, the
fewer pairs the algorithm will have to handle, e.g. for 1'000 items and a
`LinkedChunk` capacity of 128, it's only 8 pairs, be 256 bytes.

Most of the patch is documentation and tests, don't be impressed 🙂.

---

* Address https://github.com/matrix-org/matrix-rust-sdk/issues/3280

### CODE REVIEW

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs [hunk 1]
```diff
@@ -0,0 +1,529 @@
...
+        let mut updates = self.updates.write().unwrap();
+        let updates = updates.take_with_token(self.token);
+
+        self.mapper.map(updates)
+    }
+}
+
+/// Internal type that converts [`Update`] into [`VectorDiff`].
+///
+/// This type maintains its own state.
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:87]: This should be true of any data structure?

```suggestion
```

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs [hunk 2]
```diff
@@ -0,0 +1,529 @@
...
+struct UpdateToVectorDiff {
+    /// Pairs of all known chunks and their respective length. This is the only
+    /// required data for this algorithm.
+    chunks: VecDeque<(ChunkIdentifier, ChunkLength)>,
+}
+
+impl UpdateToVectorDiff {
+    /// Map several [`Update`] into [`VectorDiff`].
+    ///
+    /// How this type transforms `Update` into `VectorDiff`? There is no
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:97]:
```suggestion
    /// How does this type transform `Update` into `VectorDiff`? There is no
```

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs [hunk 3]
```diff
@@ -0,0 +1,529 @@
...
+    ///   [`VectorDiff`],
+    /// * [`Update::DetachLastItems`] is decreasing the length of the
+    ///   appropriate pair by the number of items to be detached; no
+    ///   [`VectorDiff`] is emitted,
+    /// * [`Update::ReattachItems`] and [`Update::ReattachItemsDone`] are
+    ///   respectively muting or unmuting the emission of [`VectorDiff`] by
+    ///   [`Update::PushItems`].
+    ///
+    /// The only `VectorDiff` that are emitted are [`VectorDiff::Insert`] or
+    /// [`VectorDiff::Append`] because a [`LinkedChunk`] is append-only.
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:128]: We should also be able to `Clear` a full linked chunk, in case we're seeing a limited sync (until we have a fancier diff/reconciliation algorithm)?

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs [hunk 4]
```diff
@@ -0,0 +1,529 @@
...
+    /// finds the appropriate pair (given by
+    /// [`Update::PushItems::position_hint`]). This is _the offset_. To this
+    /// offset, the algorithm adds the position's index of the new items
+    /// (still given by [`Update::PushItems::position_hint`]). This is
+    /// _the index_. This logic works for all cases as long as pairs are
+    /// maintained according to the rules hereinabove.
+    ///
+    /// That's a pretty memory compact and computation efficient way to map a
+    /// `Stream<Item = Vec<Update<Item, Gap>>>` into a `Stream<Item =
+    /// Vec<VectorDiff<Item>>>`. The larger the `LinkedChunk` capacity is, the
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:144]: nit: destination type not a `Stream` anymore

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs [hunk 5]
```diff
@@ -0,0 +1,529 @@
...
+    /// fewer pairs the algorithm will have to handle, e.g. for 1'000 items
+    /// and a `LinkedChunk` capacity of 128, it's only 8 pairs, be 256
+    /// bytes.
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:147]: what do you mean by `be` here? `i.e.` / `that is`? (if so let's use one of these two instead)

File: crates/matrix-sdk/src/event_cache/linked_chunk/updates.rs [hunk 6]
```diff
@@ -61,23 +61,31 @@ pub enum Update<Item, Gap> {
...
-    /// A chunk of kind Items has been truncated.
-    TruncateItems {
-        /// The identifier of the chunk.
-        chunk: ChunkIdentifier,
-
-        /// The new length of the chunk.
-        length: usize,
+    /// The last items of a chunk have been detached, i.e. the chunk has been
+    /// truncated.
+    DetachLastItems {
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/linked_chunk/updates.rs:81]: Not sure this name is very telling, after reading it in context. Could we rename this to `SplitOff`, maybe?

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs [hunk 7]
```diff
@@ -0,0 +1,529 @@
...
+                                .position(|(chunk_identifier, _)| chunk_identifier == next)
+                                // SAFETY: Assuming `LinkedChunk` and `Updates` are not buggy, and
+                                // assuming `Self::chunks` is correctly initialized, it is not
+                                // possible to insert a chunk between two chunks where one does not
+                                // exist. If this predicate fails, it means `LinkedChunk` or
+                                // `Updates` contain a bug.
+                                .expect("Inserting new chunk: The chunk is not found");
+
+                            debug_assert!(
+                                matches!(self.chunks.get(next_chunk_index - 1), Some((p, _)) if p == previous),
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:278]: `next_chunk_index` should always be >0 too, right? Just making sure this subtract is safe…

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs [hunk 8]
```diff
@@ -0,0 +1,529 @@
...
+                        match control_flow {
+                            ControlFlow::Break(value) => Some(value),
+                            ControlFlow::Continue(..) => None,
+                        }
+                    }
+                    // SAFETY: Assuming `LinkedChunk` and `Updates` are not buggy, and assuming
+                    // `Self::chunks` is correctly initialized, it is not possible to push items on
+                    // a chunk that does not exist. If this predicate fails, it means `LinkedChunk`
+                    // or `Updates` contain a bug.
+                    .expect("Pushing items: The chunk is not found");
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:336]: Could this be replaced with a loop? I find it hard to follow:

- we introduce an option at the end, that we try to unwrap immediately thereafter
- ending with matching that control_flow is `Continue(..)` means that we never found the item, which could be express much more explicitly (e.g. by keeping `let mut found = false` flag set to `true` when found, with an explicit `break` statement in the loop)

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs [hunk 9]
```diff
@@ -0,0 +1,529 @@
...
+
+                    *chunk_length += items.len();
+
+                    // See `mute_push_items` to learn more.
+                    if mute_push_items {
+                        continue;
+                    }
+
+                    // Optimisation: we can emit a `VectorDiff::Append` in this particular case.
+                    if chunk_index == self.chunks.len() - 1 {
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:346]: Just to make sure we don't underflow:

```suggestion
                    if chunk_index + 1 == self.chunks.len() {
```

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs [hunk 10]
```diff
@@ -0,0 +1,529 @@
...
+        // ^^^^^^^^^^^^^^^^
+        // |
+        // new
+        assert_eq!(
+            as_vector.take(),
+            &[
+                VectorDiff::Append { values: vector!['a', 'b', 'c'] },
+                VectorDiff::Append { values: vector!['d'] },
+            ]
+        );
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:424]: Nice! Alternatively, or in addition to those checks, could we also have the corresponding vector that we're building, and applying the diffs then checking its content? Then the comments don't really need to be put there, we can "just" have the assertions on the entire vectors:

```
assert_eq!(reconstructed_vector, vector!['a', 'b', 'c', 'd']);
```

File: crates/matrix-sdk/src/event_cache/linked_chunk/updates.rs [hunk 11]
```diff
@@ -61,23 +61,31 @@ pub enum Update<Item, Gap> {
...
+        /// The split position. Before this position, items are kept, after this
+        /// position, items are detached.
+        at: Position,
     },
+
+    ReattachItems,
+    ReattachItemsDone,
 }
 
 impl<Item, Gap> Clone for Update<Item, Gap>
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/updates.rs:91]: `Update<Item, Gap` implements `Clone` if and only if `Item: Clone` and `Gap: Clone`.

File: crates/matrix-sdk/src/event_cache/linked_chunk/mod.rs [hunk 12]
```diff
@@ -611,9 +729,33 @@ impl<const CAP: usize, Item, Gap> LinkedChunk<CAP, Item, Gap> {
...
     /// `Option::take()` directly but rather [`Updates::take`] for example.
+    ///
+    /// It returns `None` if updates are disabled, i.e. if this linked chunk has
+    /// been constructed with [`Self::new`], otherwise, if it's been constructed
+    /// with [`Self::new_with_update_history`], it returns `Some(…)`.
     pub fn updates(&mut self) -> Option<&mut Updates<Item, Gap>> {
         self.updates.as_mut()
     }
+
+    pub fn as_vector(&mut self) -> Option<AsVector<Item, Gap>> {
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/mod.rs:740]: How did I forget that… thanks!

File: crates/matrix-sdk/src/event_cache/linked_chunk/updates.rs [hunk 13]
```diff
@@ -61,23 +61,31 @@ pub enum Update<Item, Gap> {
...
-        /// The identifier of the chunk.
-        chunk: ChunkIdentifier,
-
-        /// The new length of the chunk.
-        length: usize,
+    /// The last items of a chunk have been detached, i.e. the chunk has been
+    /// truncated.
+    DetachLastItems {
+        /// The split position. Before this position, items are kept, after this
+        /// position, items are detached.
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/updates.rs:83]: :+1:

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs [hunk 14]
```diff
@@ -0,0 +1,529 @@
...
+    updates::{ReaderToken, Update, UpdatesInner},
+    ChunkIdentifier,
+};
+
+/// A type alias to represent a chunk's length. This is purely for commodity.
+pub(super) type ChunkLength = usize;
+
+/// A type that transforms a `Vec<Update<Item, Gap>>` —given by
+/// [`Updates`]— into a `Vec<VectorDiff<Item>>` —this
+/// type—. Basically, it helps to consume a [`LinkedChunk<CAP, Item, Gap>`] as
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:33]: :+1:

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs [hunk 15]
```diff
@@ -0,0 +1,529 @@
...
+    token: ReaderToken,
+
+    /// Mapper from `Update` to `VectorDiff`.
+    mapper: UpdateToVectorDiff,
+}
+
+impl<Item, Gap> AsVector<Item, Gap> {
+    /// Create a new [`Self`].
+    ///
+    /// `updates_subscriber` is simply `UpdatesSubscriber`.
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:52]: :+1:

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs (hunk 1 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:87]: Indeed.

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs (hunk 2 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:97]: :+1:

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs (hunk 3 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:128]: That's something I want to discuss with you. `RoomEventCacheUpdate` has a `Clear` variant already. Clearing a `LinkedChunk` means removing data from the memory and probably the database. Clearing a `RoomEventCache` means removing data from the memory only as far as I know. The meaning and the usage aren't exactly the same. Let's discuss about that and maybe let's open a new PR if necessary.

File: crates/matrix-sdk/src/event_cache/linked_chunk/updates.rs (hunk 6 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/updates.rs:81]: `Detach…` was playing well with `Reattach…` 😇.

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs (hunk 7 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:278]: Yeah correct, if `next` is at position 0, then `previous` must be at position -1 at best, which doesn't exist. Or, alternatively, `Update::NewItemsChunk` or `Update::NewGapChunk` is totally broken, but in this case, we would have other problems 😉.

This subtract is safe.

File: crates/matrix-sdk/src/event_cache/linked_chunk/updates.rs [hunk 16]
```diff
@@ -61,23 +61,31 @@ pub enum Update<Item, Gap> {
...
+    /// Items are pushed inside a chunk of kind Items.
+    PushItems {
+        /// The [`Position`] of the items.
+        ///
+        /// This value is given to prevent the need for position computations by
+        /// the update readers. Items are pushed, so the positions should be
+        /// incrementally computed from the previous items, which requires the
+        /// reading of the last previous item. With `position_hint`, the update
+        /// readers no longer need to do so.
+        position_hint: Position,
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/updates.rs:73]: I've had a hard time finding the best naming here. `at` could imply it can be _at any position_ within the chunk, so possibility before or in the middle of existing items, while it's impossible and will never happen. It's a _push_. The `Position` is not required. It's here only to save computations from the consumers. For example, if we need to insert in a database, we need to first read the items, sort them, and find the last items; from here we can compute the position for the newly pushed items. This reading + sorting can be costly, hence `position_hint`, it's a pre-computed position.

Naming it `at` sounds bad to me. Naming it `position` creates a feeling of inconsistency with the rest of the code. That's why I've decided to use `position_hint`. It may not be the best idea I had, but in my mental scheme, it works 😛.

File: crates/matrix-sdk/src/event_cache/linked_chunk/updates.rs [hunk 17]
```diff
@@ -61,23 +61,31 @@ pub enum Update<Item, Gap> {
...
+    /// The last items of a chunk have been detached, i.e. the chunk has been
+    /// truncated.
+    DetachLastItems {
+        /// The split position. Before this position, items are kept, after this
+        /// position, items are detached.
+        at: Position,
     },
+
+    ReattachItems,
+    ReattachItemsDone,
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/updates.rs:88]: Yup, good idea.

File: crates/matrix-sdk/src/event_cache/linked_chunk/updates.rs [hunk 18]
```diff
@@ -258,25 +268,37 @@ impl<Item, Gap> Updates<Item, Gap> {
...
 
-        UpdatesSubscriber { updates: Arc::downgrade(&self.inner), token }
+        AsVector::new(Arc::clone(&self.inner), token, initial_chunk_lengths)
+    }
+
+    pub(super) fn new_reader_token(&mut self) -> ReaderToken {
+        let mut inner = self.inner.write().unwrap();
+        inner.last_token += 1;
+
+        let last_token = inner.last_token;
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/updates.rs:292]: Good idea.

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs [hunk 19]
```diff
@@ -0,0 +1,529 @@
...
+    /// `initial_chunk_lengths` must be pairs of all chunk identifiers with the
+    /// associated chunk length. The pairs must be in the exact same order than
+    /// in [`LinkedChunk`].
+    ///
+    /// # Safety
+    ///
+    /// This method is marked as unsafe only to attract the caller's attention
+    /// on the order of pairs. If the order of pairs are incorrect, the entire
+    /// algorithm will not work properly and is very likely to panic.
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:61]: Indeed, it simplifies the code greatly. `initial_chunk_lengths` is now handled by `UpdateToVectorDiff::new` (🆕).

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs [hunk 20]
```diff
@@ -0,0 +1,529 @@
...
+    /// Take the new updates as [`VectorDiff`].
+    ///
+    /// It returns an empty `Vec` if there is no new `VectorDiff` for the
+    /// moment.
+    pub(super) fn take(&mut self) -> Vec<VectorDiff<Item>>
+    where
+        Item: Clone,
+    {
+        let mut updates = self.updates.write().unwrap();
+        let updates = updates.take_with_token(self.token);
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:79]: There is no different concept, I don't understand what you're referring to. I've changed the highlighted code a little bit to avoid having 2 variables with the `updates` name though.

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs (hunk 8 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:336]: I've removed the `Option`, indeed it's pointless. I wish [`ControlFlow::break_value`](https://doc.rust-lang.org/std/ops/enum.ControlFlow.html#method.break_value) lands soon in stable.

I like the `try_fold` approach better than a `for` loop though.

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs (hunk 9 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:346]: Good eye, done!

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs (hunk 10 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs]: Done.

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs (hunk 3 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:129]: Also, to have a `VectorDiff::Clear` here, we must first have an `Update::Clear`, which isn't the case so far. Let's handle that as a separate feature patch.

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs [hunk 21]
```diff
@@ -0,0 +1,573 @@
...
+                        .chunks
+                        .iter()
+                        .position(|(chunk_identifier, _)| {
+                            chunk_identifier == expected_chunk_identifier
+                        })
+                        // SAFETY: Assuming `LinkedChunk` and `Updates` are not buggy, and assuming
+                        // `Self::chunks` is correctly initialized, it is not possible to remove a
+                        // chunk that does not exist. If this predicate fails, it means
+                        // `LinkedChunk` or `Updates` contain a bug.
+                        .expect("Removing a chunk: The chunk is not found");
```
Comment (poljar) [inline crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:305]: Why is it better to crash here instead of turning this into a no-op?

**Review by poljar — APPROVED** [review]
This mostly makes sense to me, I'm not that familiar with what you're trying to achieve and don't have time to do a deep dive as of now.

I left some nits, and it would probably be wise to get approval from @bnjbvr due to the above, but feel free to move forward if you disagree.

File: crates/matrix-sdk/src/event_cache/linked_chunk/updates.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/updates.rs]: Fixed.

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs [hunk 22]
```diff
@@ -0,0 +1,573 @@
...
+use super::{
+    updates::{ReaderToken, Update, UpdatesInner},
+    ChunkContent, ChunkIdentifier, Iter,
+};
+
+/// A type alias to represent a chunk's length. This is purely for commodity.
+type ChunkLength = usize;
+
+/// A type that transforms a `Vec<Update<Item, Gap>>` (given by
+/// [`Updates`](super::Update)) into a `Vec<VectorDiff<Item>>` (this
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:32]: Oh, this doc is wrong. Fixed. Thanks!

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs [hunk 23]
```diff
@@ -0,0 +1,573 @@
...
+/// type). Basically, it helps to consume a [`LinkedChunk<CAP, Item, Gap>`] as
+/// if it was an [`eyeball::ObservableVector<Item>`].
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:34]: :+1:

File: crates/matrix-sdk/src/event_cache/linked_chunk/updates.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (bnjbvr) [reply crates/matrix-sdk/src/event_cache/linked_chunk/updates.rs]: I don't get it; if the `position_hint` is always used, and we always insert at the `position_hint` position, how is that conceptually different from the position `at` which we want to insert?

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (bnjbvr) [reply crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs]: `self.updates` is rather some kind of `UpdateManager` / `UpdateDispatcher` / `ObservableUpdates` type, but no worries, I'll make a proposal in form of a PR later :)

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs [hunk 24]
```diff
@@ -0,0 +1,573 @@
...
+
+    /// The token to read the updates.
+    token: ReaderToken,
+
+    /// Mapper from `Update` to `VectorDiff`.
+    mapper: UpdateToVectorDiff,
+}
+
+impl<Item, Gap> AsVector<Item, Gap> {
+    /// Create a new [`Self`].
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs]: Indeed, solved and improved.

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs]: Renamed to `ObservableUpdates` 👍.

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs (hunk 21 above)
Comment (poljar) [reply crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs]: Doesn't need to be silent, a log line isn't for example. But sure, if it really can't happen.

File: crates/matrix-sdk/src/event_cache/linked_chunk/updates.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/updates.rs]: Renamed to `at`.

File: crates/matrix-sdk/src/event_cache/linked_chunk/updates.rs [hunk 25]
```diff
@@ -61,75 +61,110 @@ pub enum Update<Item, Gap> {
...
-        /// The identifier of the chunk.
-        chunk: ChunkIdentifier,
-
-        /// The new length of the chunk.
-        length: usize,
+    /// The last items of a chunk have been detached, i.e. the chunk has been
+    /// truncated.
+    DetachLastItems {
+        /// The split position. Before this position (`[..position]`), items are
+        /// kept, from this position (`[position..]`), items are
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/linked_chunk/updates.rs:83]: Are the examples in the parens supposed to be using range notation? I found the `[]` confusing, and thought it was maths intervals, and I was *just* about to request either `[..position[` or `]position..]` to make this clearer. Maybe we could remove the brackets?

File: crates/matrix-sdk/src/event_cache/linked_chunk/mod.rs [hunk 26]
```diff
@@ -24,22 +102,39 @@ use std::{
...
 pub enum Error {
+    /// A chunk identifier is invalid.
     #[error("The chunk identifier is invalid: `{identifier:?}`")]
-    InvalidChunkIdentifier { identifier: ChunkIdentifier },
+    InvalidChunkIdentifier {
+        /// The chunk identifier.
+        identifier: ChunkIdentifier,
+    },
 
+    /// A chunk is a gap, and it was expected to be an items.
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/linked_chunk/mod.rs:118]:
```suggestion
    /// A chunk is a gap, and it was expected to be an items chunk.
```

File: crates/matrix-sdk/src/event_cache/linked_chunk/mod.rs [hunk 27]
```diff
@@ -24,22 +102,39 @@ use std::{
...
     #[error("The chunk is a gap: `{identifier:?}`")]
-    ChunkIsAGap { identifier: ChunkIdentifier },
+    ChunkIsAGap {
+        /// The chunk identifier.
+        identifier: ChunkIdentifier,
+    },
 
+    /// A chunk is an items, and it was expected to be a gap.
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/linked_chunk/mod.rs:125]:
```suggestion
    /// A chunk is an items chunk, and it was expected to be a gap.
```

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs [hunk 28]
```diff
@@ -0,0 +1,584 @@
...
+        drop(linked_chunk);
+        assert!(as_vector.take().is_empty());
+
+        // Finally, ensure the “reconstitued” vector is the one expected.
+        assert_eq!(
+            accumulator,
+            vector![
+                'm', 'a', 'w', 'x', 'y', 'z', 'b', 'c', 'd', 'i', 'j', 'k', 'l', 'e', 'f', 'g', 'h'
+            ]
+        );
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:582]: Instead of doing this assert once at the end, could we do it in `apply_and_assert_eq`? Iterate on all the items of the linked chunk to reconstruct a `Vector` from it, and compare it against the accumulator \o/

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs [hunk 29]
```diff
@@ -0,0 +1,584 @@
...
+
+                    // See `mute_push_items` to learn more.
+                    if mute_push_items {
+                        continue;
+                    }
+
+                    // Optimisation: we can emit a `VectorDiff::Append` in this particular case.
+                    if chunk_index + 1 == self.chunks.len() {
+                        diffs.push(VectorDiff::Append { values: items.into() });
+                    }
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:365]: Should we only emit append if `items.len() > 1`, and emit `PushBack` otherwise?

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs [hunk 30]
```diff
@@ -0,0 +1,584 @@
...
+                            // Chunk has not been found.
+                            ControlFlow::Continue(..) => {
+                                // SAFETY: Assuming `LinkedChunk` and `ObservableUpdates` are not
+                                // buggy, and assuming `Self::chunks` is correctly initialized, it
+                                // is not possible to push items on a chunk that does not exist. If
+                                // this predicate fails, it means `LinkedChunk` or
+                                // `ObservableUpdates` contain a bug.
+                                panic!("Pushing items: The chunk is not found");
+                            }
+                        }
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:352]: I still find this a bit obtuse and harder to read for newcomers to this code, would really prefer a plain loop with a `let mut found = false` flag :pensive:

**Review by bnjbvr — APPROVED** [review]
Thanks! An amazing chunk of work :drum:

File: crates/matrix-sdk/src/event_cache/linked_chunk/updates.rs (hunk 25 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/updates.rs:83]: :+1:

File: crates/matrix-sdk/src/event_cache/linked_chunk/mod.rs (hunk 27 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/mod.rs:125]: :+1:

File: crates/matrix-sdk/src/event_cache/linked_chunk/mod.rs (hunk 26 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/mod.rs:118]: :+1:

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs (hunk 28 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:582]: I think it's robust enough right now. The idea of comparing a rebuilt vector is to ensure the `VectorDiff`s we expect are the correct ones. Having this once at the end seems fine to me.

File: crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs (hunk 29 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs:365]: We can have multiple optimisations here. I would let that for another PR, this one is large enough :-D.
